frontend/components/player/player.jsx

```diff
@@ -0,0 +1,98 @@
+import React, { useEffect, useRef, useState } from "react";
+import NowPlayingInfo from "./now_playing_info";
+import PlayingControls from "./playing_controls";
+
+const Player = ({ tracks, isPlaying, reduxPlay }) => {
+	// Set local states
+	const [trackIndex, setTrackIndex] = useState(0);
+	const [trackProgress, setTrackProgress] = useState(0); // progress bar
+	const [isShuffling, setIsShuffling] = useState(false);
+	const [afterFirstPlay, setAfterFirstPlay] = useState(false);
+
+	const updateTrackProgress = (time) => {
+		return setTrackProgress(time);
+	};
+
+	// Set current track
+	let currentTrack = tracks ? tracks[trackIndex] : null;
+	let audioRef = useRef(new Audio()); // creates empty HTMLAudioElement
+	let audioSrc;
+
+	useEffect(() => {
+		audioSrc = currentTrack ? currentTrack.audioUrl : "";
+		audioRef.current.src = audioSrc;
+	}, [currentTrack]);
+
+	// Set up play/pause behavior;
+	useEffect(() => {
+		if (isPlaying) {
+			setAfterFirstPlay(true);
+			audioRef.current.play();
+		} else {
+			audioRef.current.pause();
+		}
+	}, [isPlaying]);
+	
+	// Set up behavior when changing tracks
+	const afterFirstRender = useRef(false); // prevent auto-play
+	useEffect(() => {
+		if (isPlaying && afterFirstRender.current) {
+			audioRef.current.pause();
+		}
+		audioRef.current.src = audioSrc;
+		setTrackProgress(audioRef.current.currentTime);
+		if (isPlaying) {
+			audioRef.current.play();
+		}
+		if (!afterFirstRender) afterFirstRender.current = true;
+	}, [trackIndex]);
+
+	// Create PlayingControls functions
+	const toPrevTrack = () => {
+		if (isPlaying) audioRef.current.pause();
+		if (trackIndex - 1 < 0) {
+			setTrackIndex(tracks.length - 1);
+		} else {
+			setTrackIndex(trackIndex - 1);
+		}
+	};
+	const toNextTrack = () => {
+		if (isPlaying) audioRef.current.pause();
+		if (trackIndex < tracks.length - 1) {
+			setTrackIndex(trackIndex + 1);
+		} else {
+			setTrackIndex(0);
+		}
+	};
+	const toggleShuffle = () => {
+		// TODO: Re-work this logic
+		if (isShuffling) return setIsShuffling(false);
+		if (tracks.length > 1) {
+			setIsShuffling(true);
+			setTrackIndex(Math.floor(Math.random() * tracks.length));
+		}
+	};
+
+	return (
+		<div className="player-container">
+			<NowPlayingInfo
+				audioRef={audioRef}
+				track={currentTrack}
+				trackProgress={trackProgress}
```

Comment:
Continuing [this conversation](https://www.dropbox.com/scl/fi/ubilz852ql64tcshed1a9/WIP-Leveraging-LLMs-at-Dropbox.paper?rlkey=dq3fup8y6rnrlh8uz84fwlcq5&dl=0#:uid=645710344692930422371167&h2=Proposal) here, this component seems like it's responsible for just displaying info about the current track, not really about updating the current track progress. Can we restructure the component so that it doesn't have to be aware of the `trackProgress` or be responsible for `updateTrackProgress`. Ie. let the `<Player/>` be responsible for trackProgress and when to show the `<NowPlayingInfo>`.

Reply:
This is a really really good point. This will be re-addressed after I re-configure how the NowPlaying queue is being sourced. Thank you.